paddle/fluid/operators/math/functors.h

```diff
@@ -0,0 +1,383 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <algorithm>
+#include <string>
+#include <vector>
+#include "paddle/fluid/framework/tensor.h"
+#include "paddle/fluid/framework/tensor_util.h"
+#include "paddle/fluid/operators/elementwise_op_function.h"
+#include "paddle/fluid/platform/device_context.h"
+
+namespace paddle {
+namespace operators {
+namespace math {
+
+// AddFunctor
+template <typename T>
+struct AddFunctor {
+  // out = x + y;
+  inline HOSTDEVICE T operator()(const T x, const T y) const { return x + y; }
+};
+
+template <typename T>
+struct AddGradFunctor {
+  inline HOSTDEVICE T operator()(const T x, const T y) const { return 1; }
```

Comment:
What is the difference between two `operator()`s?

Should they be two method like `DX()` and `DY()`?

Reply:
> What is the difference between two operator()s?

The operator which has two arguments doesn't use `out` to compute dx, and the operator which has three arguments can use `out` to compute dx. In `AddGradFunctor`, the implementation of the two operators is the same.

> Should they be two methods like DX() and DY()?

I found that the implementation of the two methods is the same, so is it necessary that defining them respectively?